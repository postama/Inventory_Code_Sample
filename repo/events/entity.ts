export function getEntityQuery(customerId: string, entityId: string) {
    return {
        KeyConditionExpression: 'customerId = :cust_id and begins_with(genEntityIdTime, :ent_id)',
        IndexName: 'sortedItems',
        ExpressionAttributeValues: {
            ':cust_id': customerId,
            ':ent_id': entityId
        }
    };
}