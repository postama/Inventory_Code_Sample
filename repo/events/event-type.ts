import { string, type as iotype, TypeOf, UnknownRecord, number } from 'io-ts';
import { validateType } from '../../infrastructure/validator/validator';
import { stringMatch } from '../../infrastructure/custom-brands/string-length';
import { UUID } from 'io-ts-types/lib/UUID'
import uuid from 'uuid/v4';

// These are the 'columns' of our dynamo db
export const BaseEventValidator = iotype({
    type: string,
    id: UUID,
    // The api gateway command id to ensure command is not processed twice
    commandId: string,
    version: number,
    customerId: stringMatch(6, 6),
    siteId: stringMatch(6, 6),
    user: string,
    created: string,
    entityId: UUID,
    genEntityIdTime: string,
    genEntityIdEventTypeTime: string,
    data: UnknownRecord,
    state: UnknownRecord
});

export type BaseEvent = TypeOf<typeof BaseEventValidator>;

export function createBaseEvent(
    { customerId, siteId, user, commandId },
    entityId: string,
    type: string,
    version: number): BaseEvent {
    const created = new Date();
    return validateType({
        type,
        id: uuid(),
        commandId,
        version,
        customerId,
        siteId,
        user,
        created: created.toISOString(),
        entityId,
        genEntityIdTime: `${entityId}-${created.getTime()}`,
        genEntityIdEventTypeTime: `${entityId}-${type}-${created.getTime()}`,
        data: {},
        state: {}
    }, BaseEventValidator);
}
