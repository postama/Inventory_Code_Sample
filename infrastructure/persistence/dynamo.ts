import { DocumentClient, QueryInput, PutItemInput, PutItemInputAttributeMap } from 'aws-sdk/clients/dynamodb';
import { Reducer } from '../reducer/reducer';
import { isEmpty } from '../util/hasValues';
import { BaseEvent, BaseEventValidator } from '../../repo/events/event-type';
import { validateType } from '../validator/validator';
import config from '../../appconfig.json';
const dynamoConfig = config[(process.env.NODE_CONFIG_ENV || process.env.NODE_ENV)].dynamoDB;

const dynamo = new DocumentClient(dynamoConfig.options);
const TableName = dynamoConfig.tableName;

export async function query<T>(params, options = {}): Promise<T[]> {
    const enhancedParams: QueryInput = { TableName, ...params, ...options };
    const data = await dynamo.query(enhancedParams).promise();
    return data.Items as any;
}

export async function getById<T>(params): Promise<T | null> {
    const results = await query<T>(params);
    return allowOne(results, params);
}

export function allowOne(arr, params) {
    if (!arr || !arr.length) return null;
    if (arr.length > 1) throw new Error(`Expected one result for dynamo query with params ${JSON.stringify(params)}`);
    return arr[0];
}

export async function put(item: BaseEvent): Promise<void> {
    // Final validator to make sure nothing ends up in the DB that it isn't set up for.
    validateType(item, BaseEventValidator);
    const enhancedParams: PutItemInput = { TableName, Item: item as PutItemInputAttributeMap };
    await dynamo.put(enhancedParams).promise();
    return;
}

export async function pagedReducer<T extends object, E extends { type: string }>(reducer: Reducer<T, E>, params: object, accum: T | {} = {}, options: queryOptions = { Limit: 100 }): Promise<T> {
    const enhancedParams: QueryInput = { TableName, ...params, ...options };
    const data = await dynamo.query(enhancedParams).promise();
    const items: E[] = data.Items as E[];
    const newValue = reducer.reduce(accum, items);
    if (data.LastEvaluatedKey) return pagedReducer<T, E>(reducer, params, { ...options, ExclusiveStartKey: data.LastEvaluatedKey }, newValue);
    if (!newValue || isEmpty(newValue)) return null;
    return newValue;
}

export async function pagedTask(params: object, fn: (...args: any[]) => Promise<void>, options: queryOptions = { Limit: 100 }): Promise<void> {
    const enhancedParams: QueryInput = { TableName, ...params, ...options };
    const data = await dynamo.query(enhancedParams).promise();
    const items = data.Items;
    for (const i of items) await fn(i);
    if (data.LastEvaluatedKey) return pagedTask(params, fn, { ...options, ExclusiveStartKey: data.LastEvaluatedKey });
}

type queryOptions = {
    Limit?: number,
    ExclusiveStartKey?: any
};
