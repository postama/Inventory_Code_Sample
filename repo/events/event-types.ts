import { ItemEvent } from '../items/item-event-types';
import { LocationEvent } from '../locations/location-event-types';

export type Event =
    ItemEvent |
    LocationEvent;