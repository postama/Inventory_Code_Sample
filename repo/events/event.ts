import * as dynamo from '../../infrastructure/persistence/dynamo';
import { ValidationError } from '../../../Kanso.Doorways.InventoryShared/errors';
import { Event } from './event-types';

export async function getById(customerId: string, commandId: string): Promise<Event> {
    const params = {
        KeyConditionExpression: 'customerId = :cust_id and commandId = :comm_id',
        ExpressionAttributeValues: {
            ':cust_id': customerId,
            ':comm_id': commandId
        }
    };

    return await dynamo.getById<Event>(params);
}

export async function checkUniqueCommand<T extends { customerId: string, commandId: string }>(command: T): Promise<T> {
    const event = await getById(command.customerId, command.commandId);
    if (event) throw new ValidationError(`Command ${command.commandId} has already been processed`);
    return command;
}
