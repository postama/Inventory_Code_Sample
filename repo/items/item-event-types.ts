import { ItemCreatedV1 } from '../../services/items/create-item/create-item-event';
import { ItemQtyAdjustedV1 } from '../../services/items/adjust-qty/adjust-qty-event';
import { ItemIssuedV1 } from '../../services/items/issue-item/issue-item-event';
import { ItemDisposedV1 } from '../../services/items/dispose-item/dispose-item-event';
import { ItemReceivedV1 } from '../../services/items/receive-item/receive-item-event';
import { ItemDeletedV1 } from '../../services/items/delete-item/delete-item-event';
import { ItemEditedV1 } from '../../services/items/edit-item/edit-item-event';
import { ItemTransferredV1 } from '../../services/items/transfer-item/transfer-item-event';

export type ItemEvent =
    ItemCreatedV1 |
    ItemQtyAdjustedV1 |
    ItemIssuedV1 |
    ItemDisposedV1 |
    ItemReceivedV1 |
    ItemDeletedV1 |
    ItemEditedV1 |
    ItemTransferredV1
    ;