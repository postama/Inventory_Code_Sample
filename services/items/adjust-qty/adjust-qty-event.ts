import { type, TypeOf, literal, intersection } from 'io-ts';
import { ItemActionEventValidator } from '../../../repo/items/base-item-event';
import { AdjustItemQtyCommand } from './adjust-qty-command';
import { validateType } from '../../../infrastructure/validator/validator';
import { createBaseEvent, BaseEventValidator } from '../../../repo/events/event-type';

export const ADJUST_ITEM = 'ItemQtyAdjusted';

export const ItemQtyAdjustedValidatorV1 = intersection([
    BaseEventValidator,
    type({
        type: literal(ADJUST_ITEM),
        version: literal(1),
        data: ItemActionEventValidator
    })
]);

export type ItemQtyAdjustedV1 = TypeOf<typeof ItemQtyAdjustedValidatorV1>;

export function createAdjustQtyEvent(command: AdjustItemQtyCommand): ItemQtyAdjustedV1 {
    const event = {
        ...createBaseEvent(command, command.itemId, ADJUST_ITEM, 1),
        data: {
            locationId: command.locationId,
            quantity: command.quantity,
            description: command.description
        }
    };

    return validateType(event, ItemQtyAdjustedValidatorV1);
}