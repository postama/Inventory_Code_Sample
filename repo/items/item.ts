import * as dynamo from '../../infrastructure/persistence/dynamo';
import { ItemEvent } from './item-event-types';
import { Item } from './item.type';
import { Reducer } from '../../infrastructure/reducer/reducer';
import { getEntityQuery } from '../events/entity';
import { CREATE_ITEM } from '../../services/items/create-item/create-item-event';
import { createItemReducer } from '../../services/items/create-item/create-item-reducer';
import { ADJUST_ITEM } from '../../services/items/adjust-qty/adjust-qty-event';
import { adjustQtyReducer } from '../../services/items/adjust-qty/adjust-qty-reducer';
import { ITEM_ISSUED } from '../../services/items/issue-item/issue-item-event';
import { issueItemReducer } from '../../services/items/issue-item/issue-item-reducer';
import { DISPOSE_ITEM } from '../../services/items/dispose-item/dispose-item-event';
import { disposeItemReducer } from '../../services/items/dispose-item/dispose-item-reducer';
import { RECEIVE_ITEM } from '../../services/items/receive-item/receive-item-event';
import { receiveItemReducer } from '../../services/items/receive-item/receive-item-reducer';
import { DELETE_ITEM } from '../../services/items/delete-item/delete-item-event';
import { deleteItemReducer } from '../../services/items/delete-item/delete-item-reducer';
import { EDIT_ITEM } from '../../services/items/edit-item/edit-item-event';
import { editItemReducer } from '../../services/items/edit-item/edit-item-reducer';
import { TRANSFER_ITEM } from '../../services/items/transfer-item/transfer-item-event';
import { transferItemReducer } from '../../services/items/transfer-item/transfer-item-reducer';

export const ItemReducer = new Reducer<Item, ItemEvent>();

// Code smell - breaking Open/Closed Principal these should be registered by their reducers, not in a central file.
// Not sure how to fix this - they need to be registered for every code split lambda, but how will each code split lambda bring in the code from the other events?
ItemReducer.register(CREATE_ITEM, createItemReducer);
ItemReducer.register(ADJUST_ITEM, adjustQtyReducer);
ItemReducer.register(ITEM_ISSUED, issueItemReducer);
ItemReducer.register(DISPOSE_ITEM, disposeItemReducer);
ItemReducer.register(RECEIVE_ITEM, receiveItemReducer);
ItemReducer.register(DELETE_ITEM, deleteItemReducer);
ItemReducer.register(EDIT_ITEM, editItemReducer);
ItemReducer.register(TRANSFER_ITEM, transferItemReducer);

export async function getById(customerId: string, itemId: string): Promise<Item> {
    return await dynamo.pagedReducer<Item, ItemEvent>(ItemReducer, getEntityQuery(customerId, itemId));
}

export async function save(event: ItemEvent) {
    return await dynamo.put(event);
}
