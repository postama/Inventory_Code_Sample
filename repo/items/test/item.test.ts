import { ItemReducer, getById, save } from '../item';
import uuid from 'uuid/v4';
import { Item } from '../item.type';
import { pagedReducer } from '../../../infrastructure/persistence/dynamo';
import { ItemEvent } from '../item-event-types';
import { getEntityQuery } from '../../events/entity';
import { createTestCreateItemEvent } from '../../../services/items/create-item/tests/test.util';
import { createTestAdjustEvent } from '../../../services/items/adjust-qty/tests/test.util';
import { ItemCreatedV1 } from '../../../services/items/create-item/create-item-event';
import { ItemQtyAdjustedV1 } from '../../../services/items/adjust-qty/adjust-qty-event';

describe('inventory aggregate root', () => {
    describe('event reducer', () => {
        test('should reduce events to an inventory item', async () => {
            const event1: ItemCreatedV1 = createTestCreateItemEvent();
            const event2: ItemQtyAdjustedV1 = createTestAdjustEvent({ customerId: event1.customerId, itemId: event1.entityId }, {});
            const event3: ItemQtyAdjustedV1 = createTestAdjustEvent({ customerId: event1.customerId, itemId: event1.entityId }, {});
            const item = ItemReducer.reduce({}, [event1, event2, event3]);
            expect(item.quantity).toBe(2);
            expect(item.created).toBe(event1.created);
        });

        test('should throw if create does not come first', async () => {
            try {
                const itemId = uuid();
                const event1: ItemQtyAdjustedV1 = createTestAdjustEvent({ itemId }, {});
                const event2: ItemCreatedV1 = createTestCreateItemEvent({ customerId: event1.customerId }, { itemId });
                ItemReducer.reduce({}, [event1, event2]);
            } catch (e) {
                expect(e.message).toBe('Item does not exist, event cannot be processed');
            }
        });

        test('should work across multiple pages of events', async () => {
            const event: ItemCreatedV1 = createTestCreateItemEvent();
            await save(event);
            const adjust1: ItemQtyAdjustedV1 = createTestAdjustEvent({ customerId: event.customerId, itemId: event.entityId });
            const adjust2: ItemQtyAdjustedV1 = createTestAdjustEvent({ customerId: event.customerId, itemId: event.entityId });
            await Promise.all([save(adjust1), save(adjust2)]);

            const query = getEntityQuery(event.customerId, event.entityId);
            const result = await pagedReducer<Item, ItemEvent>(ItemReducer, query, undefined, { Limit: 1 });
            expect(result.quantity).toBe(2);
        });

        test('should return null if no events', async () => {
            const result = await pagedReducer(ItemReducer, getEntityQuery(uuid(), uuid()), undefined, { Limit: 1 });
            expect(result).toBeNull();
        });
    });

    describe('getById', () => {
        test('should get an inventory item', async () => {
            const event: ItemCreatedV1 = createTestCreateItemEvent();
            await save(event);
            const item = await getById(event.customerId, event.entityId);
            expect(item).toMatchSnapshot({
                created: expect.any(String)
            });
        });
    });
});