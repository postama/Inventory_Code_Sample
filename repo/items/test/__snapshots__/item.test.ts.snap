// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inventory aggregate root getById should get an inventory item 1`] = `
Object {
  "brand": "abc",
  "costPerUnit": 123,
  "created": Any<String>,
  "group": "abc",
  "identifier": "abc",
  "isPreferred": true,
  "locations": Object {},
  "name": "abc",
  "quantity": 0,
  "supplier": "abc",
  "unitOfMeasurement": "abc",
}
`;
